### PR TITLE
Add ResidualAccumulator

### LINKED ISSUE #16: Add a utility for accumulating point data into the residual
## Reason
We want to accumulate the charge field into a residual as a source.

## Design
Add a `ResidualAccumulator` with an interface similar to the `AuxAccumulator` for dumping point sources.

## Impact
New feature

### PR DESCRIPTION
**Pull request by loganharbour**
Closes #16

### CODE REVIEW

File: doc/content/source/utils/ResidualAccumulator.md [hunk 1]
```diff
@@ -0,0 +1,16 @@
+# ResidualAccumulator
+
+The `ResidualAccumulator` is a utility that enables the accumulation of pointwise data
+into a residual as a point source. The accumulation is done similarly to how Dirac points
+would contribute to a field.
+
+It is designed to be constructed and destructed on use. That is - you should
+instantiate an `ResidualAccumulator` when you want to contribute to the field, and then
```
Comment (gsgall) [inline doc/content/source/utils/ResidualAccumulator.md]:
```suggestion
instantiate a `ResidualAccumulator` when you want to contribute to the field, and then
```

File: doc/content/source/utils/ResidualAccumulator.md [hunk 2]
```diff
@@ -0,0 +1,16 @@
...
+call `ResidualAccumulator::finalize()` and destruct it once you are done contributing
+to the field. The finalization will take care of properly closing the aux variable.
```
Comment (gsgall) [inline doc/content/source/utils/ResidualAccumulator.md]: Doe the residual accumulator operate only on aux variables. I thought this required a nonLinearVariable to work properly.

File: include/utils/ResidualAccumulator.h [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+                      const unsigned int nl_sys_num);
+
+  virtual void add(const Elem & elem, const Point & point, const Real & value) override final;
+
+private:
+  void addCachedValues() override final;
+  void initCachedValues() override final;
+
+  /// The AuxVariable this AuxRayKernel contributes to
+  const MooseVariableField<Real> & _var;
```
Comment (gsgall) [inline include/utils/ResidualAccumulator.h]: Is this comment accurate? I didn't think that the residual accumulator required a RayKernel.

**Review by gsgall — CHANGES_REQUESTED** [review]

File: include/utils/AccumulatorBase.h [hunk 4]
```diff
@@ -0,0 +1,77 @@
...
+#pragma once
+
+#include "MooseTypes.h"
+
+class FEProblemBase;
+
+namespace FENIX
+{
+/**
+ * Base class for a uility that accumulates point values into a variable
```
Comment (cticenhour) [inline include/utils/AccumulatorBase.h]:
```suggestion
 * Base class for a utility that accumulates point values into a variable
```

**Review by cticenhour — CHANGES_REQUESTED** [review]
Noticed one small thing. Requesting changes until @gsgall's comments are addressed.

**Comment by loganharbour** [comment]
Should be good to go now.

**Review by gsgall — APPROVED** [review]
Looks good to me thank you Logan!

**Review by cticenhour — APPROVED** [review]

**Comment by cticenhour** [comment]
@loganharbour Unfortunately, this isn't Valgrind clean: https://civet.inl.gov/job/2087206/

I may be able to take a peek today.

**Comment by loganharbour** [comment]
> @loganharbour Unfortunately, this isn't Valgrind clean: https://civet.inl.gov/job/2087206/
> 
> I may be able to take a peek today.

Resolved in https://github.com/idaholab/fenix/pull/18

**Pull request merged**